### ISSUE TITLE
Ubuntu 20.04 "error: invalid command 'bdist_wheel'"

### ISSUE THREAD

**Issue by andoresuperesu** [lektor/lektor#784]
I followed the procedure on Lektor's webpage, and despite having installed all prerequisites, I get the following errors:

````console
Building wheel for inifile (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/local/lib/lektor/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-af6y5v9m/inifile/setup.py'"'"'; __file__='"'"'/tmp/pip-install-af6y5v9m/inifile/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-ayl9hhi2
       cwd: /tmp/pip-install-af6y5v9m/inifile/
  Complete output (6 lines):
  usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
     or: setup.py --help [cmd1 cmd2 ...]
     or: setup.py --help-commands
     or: setup.py cmd --help

  error: invalid command 'bdist_wheel'
  ----------------------------------------
  ERROR: Failed building wheel for inifile
  Running setup.py clean for inifile
  Building wheel for EXIFRead (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/local/lib/lektor/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-af6y5v9m/EXIFRead/setup.py'"'"'; __file__='"'"'/tmp/pip-install-af6y5v9m/EXIFRead/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-mxtpmzko
       cwd: /tmp/pip-install-af6y5v9m/EXIFRead/
  Complete output (7 lines):
  Warning: 'classifiers' should be a list, got type 'tuple'
  usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
     or: setup.py --help [cmd1 cmd2 ...]
     or: setup.py --help-commands
     or: setup.py cmd --help

  error: invalid command 'bdist_wheel'
  ----------------------------------------
  ERROR: Failed building wheel for EXIFRead
  Running setup.py clean for EXIFRead
  Building wheel for watchdog (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/local/lib/lektor/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-af6y5v9m/watchdog/setup.py'"'"'; __file__='"'"'/tmp/pip-install-af6y5v9m/watchdog/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-5pxag3tc
       cwd: /tmp/pip-install-af6y5v9m/watchdog/
  Complete output (6 lines):
  usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
     or: setup.py --help [cmd1 cmd2 ...]
     or: setup.py --help-commands
     or: setup.py cmd --help

  error: invalid command 'bdist_wheel'
  ----------------------------------------
  ERROR: Failed building wheel for watchdog
  Running setup.py clean for watchdog
  Building wheel for pathtools (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/local/lib/lektor/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-af6y5v9m/pathtools/setup.py'"'"'; __file__='"'"'/tmp/pip-install-af6y5v9m/pathtools/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-xurhnhnp
       cwd: /tmp/pip-install-af6y5v9m/pathtools/
  Complete output (8 lines):
  /tmp/pip-install-af6y5v9m/pathtools/setup.py:25: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp
  usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
     or: setup.py --help [cmd1 cmd2 ...]
     or: setup.py --help-commands
     or: setup.py cmd --help

  error: invalid command 'bdist_wheel'
  ----------------------------------------
  ERROR: Failed building wheel for pathtools
````
Any sugestions? It installed without issue in windows, so I suppose it is something related to Ubuntu 20.04.

**Comment by runfalk** [comment]
`bdist_wheel` is made available by doing `pip install wheel`. I think it is not part of standard setuptools or distutils.

Which instructions did you follow? Dev?

**Comment by andoresuperesu** [comment]
For some reason, installing each package separately before running  (therefore building them) then running again the installer worked. The question then is: why isn't it possible while running the setup script?

That resource seems to be unailable while running the install script. Wheel comes preloaded in ubuntu20.04 as on install it prompted that is was already installed. This I've tried on 2 instances already, a vps and wsl2 on windows. Both on ubuntu 20.04.

**Comment by andoresuperesu** [comment]
> `bdist_wheel` is made available by doing `pip install wheel`. I think it is not part of standard setuptools or distutils.
> 
> Which instructions did you follow? Dev?

CLI: https://www.getlektor.com/docs/installation/

**Comment by runfalk** [comment]
There are multiple installation instructions there, but I assume curl installer.py then?

I find it a bit odd that it tries to make wheels I would expect that it should just download prebuilt ones.

**Comment by xlotlu** [comment]
@andresperezcera can you post the entire output? I suspect the wheel-build failures are in fact not the reason -- you can see how it continues for each package without an early bailout.

**Comment by andoresuperesu** [comment]
@xlotlu Sorry, solved it as mentioned, however I need practice at the vps instance, I'll reset and proceed again. I'll be back with results in about 4 hours (currently busy)

**Comment by andoresuperesu** [comment]
Sorry this took forever, a hard week. Here is the full transcript on a new install of Ubuntu 20.04:

First the error, then me installing the error generating packages without issue. After installing those packages, Lektor install works without issue. 

```` console
aperez@aperez-V:~$ curl -sf https://www.getlektor.com/installer.py | sudo python3

Welcome to Lektor
This script will install Lektor on your computer.

Installing at:
  bin: /usr/local/bin
  app: /usr/local/lib/lektor

An existing installation was detected. This will be removed!

Continue? [Yn] y

Installing virtual environment...
Collecting Lektor
  Downloading Lektor-3.1.3-py2.py3-none-any.whl (1.8 MB)
     |████████████████████████████████| 1.8 MB 1.3 MB/s 
Collecting inifile
  Downloading inifile-0.4.zip (7.4 kB)
Requirement already satisfied, skipping upgrade: setuptools in /usr/local/lib/lektor/lib/python3.8/site-packages (from Lektor) (44.0.0)
Collecting EXIFRead
  Downloading ExifRead-2.1.2.tar.gz (33 kB)
Collecting Jinja2>=2.4
  Downloading Jinja2-2.11.2-py2.py3-none-any.whl (125 kB)
     |████████████████████████████████| 125 kB 16.3 MB/s 
Collecting Babel
  Downloading Babel-2.8.0-py2.py3-none-any.whl (8.6 MB)
     |████████████████████████████████| 8.6 MB 9.6 MB/s 
Collecting Flask
  Downloading Flask-1.1.2-py2.py3-none-any.whl (94 kB)
     |████████████████████████████████| 94 kB 525 kB/s 
Requirement already satisfied, skipping upgrade: pip in /usr/local/lib/lektor/lib/python3.8/site-packages (from Lektor) (20.0.2)
Collecting watchdog
  Downloading watchdog-0.10.2.tar.gz (95 kB)
     |████████████████████████████████| 95 kB 772 kB/s 
Collecting click>=6.0
  Downloading click-7.1.2-py2.py3-none-any.whl (82 kB)
     |████████████████████████████████| 82 kB 163 kB/s 
Collecting requests[security]
  Downloading requests-2.23.0-py2.py3-none-any.whl (58 kB)
     |████████████████████████████████| 58 kB 933 kB/s 
Collecting mistune>=0.7.0
  Downloading mistune-0.8.4-py2.py3-none-any.whl (16 kB)
Collecting MarkupSafe>=0.23
  Downloading MarkupSafe-1.1.1-cp38-cp38-manylinux1_x86_64.whl (32 kB)
Collecting pytz>=2015.7
  Downloading pytz-2020.1-py2.py3-none-any.whl (510 kB)
     |████████████████████████████████| 510 kB 12.5 MB/s 
Collecting itsdangerous>=0.24
  Downloading itsdangerous-1.1.0-py2.py3-none-any.whl (16 kB)
Collecting Werkzeug>=0.15
  Downloading Werkzeug-1.0.1-py2.py3-none-any.whl (298 kB)
     |████████████████████████████████| 298 kB 12.9 MB/s 
Collecting pathtools>=0.1.1
  Downloading pathtools-0.1.2.tar.gz (11 kB)
Collecting urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1
  Downloading urllib3-1.25.9-py2.py3-none-any.whl (126 kB)
     |████████████████████████████████| 126 kB 11.3 MB/s 
Collecting chardet<4,>=3.0.2
  Downloading chardet-3.0.4-py2.py3-none-any.whl (133 kB)
     |████████████████████████████████| 133 kB 15.6 MB/s 
Collecting idna<3,>=2.5
  Downloading idna-2.9-py2.py3-none-any.whl (58 kB)
     |████████████████████████████████| 58 kB 881 kB/s 
Collecting certifi>=2017.4.17
  Downloading certifi-2020.4.5.1-py2.py3-none-any.whl (157 kB)
     |████████████████████████████████| 157 kB 11.9 MB/s 
Collecting pyOpenSSL>=0.14; extra == "security"
  Downloading pyOpenSSL-19.1.0-py2.py3-none-any.whl (53 kB)
     |████████████████████████████████| 53 kB 356 kB/s 
Collecting cryptography>=1.3.4; extra == "security"
  Downloading cryptography-2.9.2-cp35-abi3-manylinux2010_x86_64.whl (2.7 MB)
     |████████████████████████████████| 2.7 MB 12.1 MB/s 
Collecting six>=1.5.2
  Downloading six-1.15.0-py2.py3-none-any.whl (10 kB)
Collecting cffi!=1.11.3,>=1.8
  Downloading cffi-1.14.0-cp38-cp38-manylinux1_x86_64.whl (409 kB)
     |████████████████████████████████| 409 kB 11.6 MB/s 
Collecting pycparser
  Downloading pycparser-2.20-py2.py3-none-any.whl (112 kB)
     |████████████████████████████████| 112 kB 17.2 MB/s 
Building wheels for collected packages: inifile, EXIFRead, watchdog, pathtools
  Building wheel for inifile (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/local/lib/lektor/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-ezhn8uhu/inifile/setup.py'"'"'; __file__='"'"'/tmp/pip-install-ezhn8uhu/inifile/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-ma23lgdd
       cwd: /tmp/pip-install-ezhn8uhu/inifile/
  Complete output (6 lines):
  usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
     or: setup.py --help [cmd1 cmd2 ...]
     or: setup.py --help-commands
     or: setup.py cmd --help
  
  error: invalid command 'bdist_wheel'
  ----------------------------------------
  ERROR: Failed building wheel for inifile
  Running setup.py clean for inifile
  Building wheel for EXIFRead (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/local/lib/lektor/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-ezhn8uhu/EXIFRead/setup.py'"'"'; __file__='"'"'/tmp/pip-install-ezhn8uhu/EXIFRead/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-_qfarw3o
       cwd: /tmp/pip-install-ezhn8uhu/EXIFRead/
  Complete output (7 lines):
  Warning: 'classifiers' should be a list, got type 'tuple'
  usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
     or: setup.py --help [cmd1 cmd2 ...]
     or: setup.py --help-commands
     or: setup.py cmd --help
  
  error: invalid command 'bdist_wheel'
  ----------------------------------------
  ERROR: Failed building wheel for EXIFRead
  Running setup.py clean for EXIFRead
  Building wheel for watchdog (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/local/lib/lektor/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-ezhn8uhu/watchdog/setup.py'"'"'; __file__='"'"'/tmp/pip-install-ezhn8uhu/watchdog/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-0ztm0kd5
       cwd: /tmp/pip-install-ezhn8uhu/watchdog/
  Complete output (6 lines):
  usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
     or: setup.py --help [cmd1 cmd2 ...]
     or: setup.py --help-commands
     or: setup.py cmd --help
  
  error: invalid command 'bdist_wheel'
  ----------------------------------------
  ERROR: Failed building wheel for watchdog
  Running setup.py clean for watchdog
  Building wheel for pathtools (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/local/lib/lektor/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-ezhn8uhu/pathtools/setup.py'"'"'; __file__='"'"'/tmp/pip-install-ezhn8uhu/pathtools/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-va118rhf
       cwd: /tmp/pip-install-ezhn8uhu/pathtools/
  Complete output (8 lines):
  /tmp/pip-install-ezhn8uhu/pathtools/setup.py:25: DeprecationWarning: the imp module is deprecated in favour of importlib; see the modules documentation for alternative uses
    import imp
  usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
     or: setup.py --help [cmd1 cmd2 ...]
     or: setup.py --help-commands
     or: setup.py cmd --help
  
  error: invalid command 'bdist_wheel'
  ----------------------------------------
  ERROR: Failed building wheel for pathtools
  Running setup.py clean for pathtools
Failed to build inifile EXIFRead watchdog pathtools
Installing collected packages: inifile, EXIFRead, MarkupSafe, Jinja2, pytz, Babel, click, itsdangerous, Werkzeug, Flask, pathtools, watchdog, urllib3, chardet, idna, certifi, pycparser, cffi, six, cryptography, pyOpenSSL, requests, mistune, Lektor
    Running setup.py install for inifile ... done
    Running setup.py install for EXIFRead ... done
    Running setup.py install for pathtools ... done
    Running setup.py install for watchdog ... done
Successfully installed Babel-2.8.0 EXIFRead-2.1.2 Flask-1.1.2 Jinja2-2.11.2 Lektor-3.1.3 MarkupSafe-1.1.1 Werkzeug-1.0.1 certifi-2020.4.5.1 cffi-1.14.0 chardet-3.0.4 click-7.1.2 cryptography-2.9.2 idna-2.9 inifile-0.4 itsdangerous-1.1.0 mistune-0.8.4 pathtools-0.1.2 pyOpenSSL-19.1.0 pycparser-2.20 pytz-2020.1 requests-2.23.0 six-1.15.0 urllib3-1.25.9 watchdog-0.10.2

All done!

aperez@aperez-V:~$ sudo pip3 install inifile
Collecting inifile
  Using cached inifile-0.4.zip (7.4 kB)
Building wheels for collected packages: inifile
  Building wheel for inifile (setup.py) ... done
  Created wheel for inifile: filename=inifile-0.4-py3-none-any.whl size=6809 sha256=157fc596fe32ddc20ab64041198dd3185f7047ec902245ae7319b0a1a22c6a6f
  Stored in directory: /root/.cache/pip/wheels/df/9c/4f/a21e9e01f6d37f98e61fe3e13f9a4a5274a9162b6d521b9cd5
Successfully built inifile
Installing collected packages: inifile
Successfully installed inifile-0.4
aperez@aperez-V:~$ sudo pip3 install EXIFRead
Collecting EXIFRead
  Using cached ExifRead-2.1.2.tar.gz (33 kB)
Building wheels for collected packages: EXIFRead
  Building wheel for EXIFRead (setup.py) ... done
  Created wheel for EXIFRead: filename=ExifRead-2.1.2-py3-none-any.whl size=34200 sha256=56cb165cc419eddf742da9c43d00ccc63f587b651fa4eb8e5f8f0f787291b37a
  Stored in directory: /root/.cache/pip/wheels/34/c4/f1/0c1e430b02eb990a0e0773b3791ba601fece63cebb836ba780
Successfully built EXIFRead
Installing collected packages: EXIFRead
Successfully installed EXIFRead-2.1.2
aperez@aperez-V:~$ sudo pip3 install watchdog
Collecting watchdog
  Using cached watchdog-0.10.2.tar.gz (95 kB)
Collecting pathtools>=0.1.1
  Using cached pathtools-0.1.2.tar.gz (11 kB)
Building wheels for collected packages: watchdog, pathtools
  Building wheel for watchdog (setup.py) ... done
  Created wheel for watchdog: filename=watchdog-0.10.2-py3-none-any.whl size=73602 sha256=d8582b445d4ad6b86be44da2099ca4403998a51b062c4c2c2e1a396fb5cb3320
  Stored in directory: /root/.cache/pip/wheels/4e/14/01/1676612f5b23c641868de61d36ac2d5be12f4eec6c087fbd62
  Building wheel for pathtools (setup.py) ... done
  Created wheel for pathtools: filename=pathtools-0.1.2-py3-none-any.whl size=8784 sha256=87fe2590c4c46809b4eb438932d352dd6c898d57d7f4c19e7e7a7e09584bcfd7
  Stored in directory: /root/.cache/pip/wheels/4c/8e/7e/72fbc243e1aeecae64a96875432e70d4e92f3d2d18123be004
Successfully built watchdog pathtools
Installing collected packages: pathtools, watchdog
Successfully installed pathtools-0.1.2 watchdog-0.10.2
````
Any comments?

**Comment by xlotlu** [comment]
> Any comments?

Well, the install is successful, as shown by the `All done!` line. The wheel failures are "only warnings".

Now, why pip should decide to build wheels when the wheel package is not available, I have no idea. It works outside the virtualenv because you have the python3-wheel package installed at system level.

Maybe we should default to `--prefer-binary` for pip install? @runfalk what do you think? We already do it on windows.

**Comment by runfalk** [comment]
I don't know enough about this to have a strong opinion. I had no idea pip would try to build wheels before installing. Shouldn't this be fixable by installing `wheel` into the virtualenv? We could update setuptools and install the wheel package in one go `pip install --upgrade pip setuptools wheel`?

**Comment by xlotlu** [comment]
Update: `--prefer-binary` won't do anything, there are no binaries to get.

This seems to me entirely very silly. The issue here is that those packages don't have wheels pre-built for py3 (see https://pypi.org/simple/exifread/ , https://pypi.org/simple/inifile/ etc.) *and* that pip tries to build a wheel locally, regardless. And funnily enough, there is in fact no issue, since everything just works.

I personally consider this a pip bug, but I'm not willing to pursue that stance. :) This issue here seems repeated for many projects, and consistently people propose `pip install wheel`, and nobody bothers to ask "but why?".

>  We could update setuptools and install the wheel package in one go `pip install --upgrade pip setuptools wheel`?

Well, "there's nothing to fix" but it saves us some tilting at windmills, so ok.

**Comment by talalash** [comment]
sorry for digging up an old topic. But I just wanted to try out lektor on Ubuntu 20.04.2 LTS, and I got exactly the same error. Found this topic here, read through it. Didn't find a solution in it. I read "it's just warnings because it said 'all done'", but when I try to do "lektor quickstart", it doesn't work either. Just leaving this note here, maybe you want to look at it.

Thought I'd give the windows version a try. In a cmd, entered "python" which let me install python 3.9.5 from the store. Then pasted the command from your site, and got an error as well. No installation. I guess Lektor is more geared towards developers.

**Comment by andoresuperesu** [comment]
@talalash the only way I found was to first install the packages and then run lektor's installer script. At the moment, though, there is a compatibility issue between Werkzeug and flask versions. That should be resolved on #911 

However windows install should run mostly without issue (wait, there's the same compatibility issue at the moment). Yet, if python is installed through the Windows Store python works a little differently and blocks the installer as is. There is an issue on this #798.

**Comment by andoresuperesu** [comment]
I'm updating this topic to point that the best install method at the moment is using pipx:

```console
$ python3 -m pip install --user pipx
$ python3 -m pipx ensurepath
```

Close and reopen terminal

```console
$ pipx install lektor
```
I'll close this issue with this and point out 3 things, 2 and 3 are derived from 1:
- First, the install script should be set aside to prioritize pipx.
- Second, since pipx won't do system-wide, we have to determine wether that's important.
- Third, since pipx only workd on 3.6+ we should decide whether all previous version are deprecated.

For further discussion please go to #915
